apiserver/facades/controller/charmrevisionupdater/updater.go

```diff
@@ -101,89 +106,205 @@ var NewCharmStoreClient = func(st *state.State) (charmstore.Client, error) {
-		"controller_version": version.Current.String(),
-		"cloud":              model.CloudName(),
-		"cloud_region":       model.CloudRegion(),
-		"is_controller":      strconv.FormatBool(model.IsControllerModel()),
+	return latest, nil
+}
+
+func fetchCharmstoreInfos(st *state.State, ids []charmstore.CharmID, apps []*state.Application) ([]latestCharmInfo, error) {
+	client, err := NewCharmStoreClient(st)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	cloud, err := st.Cloud(model.CloudName())
+	metadata, err := charmstoreAPIMetadata(st)
 	if err != nil {
-		metadata["provider"] = "unknown"
-	} else {
-		metadata["provider"] = cloud.Type
+		return nil, errors.Trace(err)
 	}
-	results, err := charmstore.LatestCharmInfo(client, charms, metadata)
+	results, err := charmstore.LatestCharmInfo(client, ids, metadata)
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
 
 	var latest []latestCharmInfo
 	for i, result := range results {
+		if i >= len(apps) {
+			logger.Errorf("retrieved more results (%d) than charmstore applications (%d)",
+				i, len(apps))
+			break
+		}
 		if result.Error != nil {
-			logger.Errorf("retrieving charm info for %s: %v", charms[i].URL, result.Error)
+			logger.Errorf("retrieving charm info for %s: %v", ids[i].URL, result.Error)
+			continue
+		}
+		application := apps[i]
+		latest = append(latest, latestCharmInfo{
+			url:         result.CharmInfo.OriginalURL.WithRevision(result.CharmInfo.LatestRevision),
+			timestamp:   result.CharmInfo.Timestamp,
+			revision:    result.CharmInfo.LatestRevision,
+			resources:   result.CharmInfo.LatestResources,
+			application: application,
+		})
+	}
+	return latest, nil
+}
+
+func fetchCharmhubInfos(st *state.State, ids []charmhubID, apps []*state.Application) ([]latestCharmInfo, error) {
+	client, err := NewCharmhubClient(st)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	results, err := charmhubLatestCharmInfo(client, ids)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var latest []latestCharmInfo
+	for i, result := range results {
+		if i >= len(apps) {
+			logger.Errorf("retrieved more results (%d) than charmhub applications (%d)",
+				i, len(apps))
+			break
+		}
+		if result.error != nil {
+			logger.Errorf("retrieving charm info for ID %s: %v", ids[i].id, result.error)
 			continue
 		}
-		application := resultsIndexedApps[i]
+		application := apps[i]
+		url, _ := application.CharmURL()
 		latest = append(latest, latestCharmInfo{
-			CharmInfo:   result.CharmInfo,
+			url:         url.WithRevision(result.revision),
+			timestamp:   result.timestamp,
+			revision:    result.revision,
+			resources:   result.resources,
 			application: application,
 		})
 	}
 	return latest, nil
 }
 
+// charmstoreAPIMetadata returns a map containing metadata key/value pairs to
+// send to the charmstore API for tracking metrics.
+func charmstoreAPIMetadata(st *state.State) (map[string]string, error) {
```

Comment:
We still want to send this for charmhub as well. They will eventually use it.
We can drop the "environment_uuid" value.